File: authportal/src/main/java/com/selva/authportal/service/AuthService.java
package com.selva.authportal.service;

import com.selva.authportal.dto.AuthResponse;
import com.selva.authportal.dto.LoginRequest;
import com.selva.authportal.dto.RegisterRequest;
import com.selva.authportal.dto.UserResponse;
import com.selva.authportal.exception.InvalidCredentialsException;
import com.selva.authportal.exception.ResourceNotFoundException;
import com.selva.authportal.exception.UserAlreadyExistsException;
import com.selva.authportal.exception.UserDisabledException;
import com.selva.authportal.model.AuthProvider;
import com.selva.authportal.model.Role;
import com.selva.authportal.model.User;
import com.selva.authportal.repository.UserRepository;
import com.selva.authportal.security.JwtService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import com.selva.authportal.security.oauth2.OAuth2ExchangeCodeStore;
import org.springframework.transaction.annotation.Transactional;

/**
 * Service orchestrating user registration, local authentication, and session token issuance.
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class AuthService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final EmailRoleResolver emailRoleResolver;
    private final JwtService jwtService;
    private final OAuth2ExchangeCodeStore exchangeCodeStore;

    /**
     * Registers a new user with local email/password credentials.
     * The role is strictly determined server-side from the institutional email.
     */
    @Transactional
    public AuthResponse register(RegisterRequest request) {
        String normalizedEmail = request.getEmail().trim().toLowerCase();

        // 1. Determine role server-side (throws InvalidInstitutionalEmailException if invalid)
        Role assignedRole = emailRoleResolver.resolveRole(normalizedEmail);

        // 2. Check for duplicate account
        if (userRepository.existsByEmail(normalizedEmail)) {
            throw new UserAlreadyExistsException("An account with this email address already exists.");
        }

        // 3. Hash password
        String hashedPassword = passwordEncoder.encode(request.getPassword());

        // 4. Create and persist user entity
        User user = User.builder()
                .name(request.getName().trim())
                .email(normalizedEmail)
                .password(hashedPassword)
                .role(assignedRole)
                .authProvider(AuthProvider.LOCAL)
                .enabled(true)
                .build();

        User savedUser = userRepository.save(user);
        log.info("Successfully registered local user with email: {} and role: {}", normalizedEmail, assignedRole);

        // 5. Generate application JWT
        String token = jwtService.generateToken(savedUser);

        return AuthResponse.builder()
                .token(token)
                .tokenType("Bearer")
                .expiresIn(jwtService.getExpirationMs() / 1000)
                .user(UserResponse.fromEntity(savedUser))
                .build();
    }

    /**
     * Authenticates a user using local email and password.
     */
    @Transactional(readOnly = true)
    public AuthResponse login(LoginRequest request) {
        String normalizedEmail = request.getEmail().trim().toLowerCase();

        // 1. Locate user; use generic error to prevent account enumeration
        User user = userRepository.findByEmail(normalizedEmail)
                .orElseThrow(() -> new InvalidCredentialsException("Invalid email or password."));

        // 2. Verify password (accounts created solely via Google OAuth will have null password)
        if (user.getPassword() == null || !passwordEncoder.matches(request.getPassword(), user.getPassword())) {
            throw new InvalidCredentialsException("Invalid email or password.");
        }

        // 3. Check account enabled status
        if (!user.isEnabled()) {
            throw new UserDisabledException("Your account is disabled. Please contact administration.");
        }

        log.info("User logged in successfully: {}", normalizedEmail);

        // 4. Generate JWT
        String token = jwtService.generateToken(user);

        return AuthResponse.builder()
                .token(token)
                .tokenType("Bearer")
                .expiresIn(jwtService.getExpirationMs() / 1000)
                .user(UserResponse.fromEntity(user))
                .build();
    }

    /**
     * Retrieves profile information for the authenticated user.
     */
    @Transactional(readOnly = true)
    public UserResponse getCurrentUser(String email) {
        User user = userRepository.findByEmail(email.trim().toLowerCase())
                .orElseThrow(() -> new ResourceNotFoundException("User not found for email: " + email));
        return UserResponse.fromEntity(user);
    }

    /**
     * Exchanges a short-lived, single-use authorization code for an application JWT.
     * Burns the code upon exchange and returns standard AuthResponse.
     */
    @Transactional(readOnly = true)
    public AuthResponse exchangeOAuthCode(String code) {
        String email = exchangeCodeStore.consumeCode(code)
                .orElseThrow(() -> new InvalidCredentialsException("Invalid or expired OAuth authorization code."));

        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new ResourceNotFoundException("User not found for email: " + email));

        if (!user.isEnabled()) {
            throw new UserDisabledException("Your account is disabled. Please contact administration.");
        }

        String token = jwtService.generateToken(user);
        log.info("Successfully exchanged one-time code for JWT for user: {}", email);

        return AuthResponse.builder()
                .token(token)
                .tokenType("Bearer")
                .expiresIn(jwtService.getExpirationMs() / 1000)
                .user(UserResponse.fromEntity(user))
                .build();
    }
}

File: frontend/src/services/authService.js
import { API_ENDPOINTS } from '../config/api';

const TOKEN_KEY = 'auth_token';
const USER_KEY = 'auth_user';

/**
 * Authentication service handling communication with Spring Boot backend endpoints.
 */
export const authService = {
  /**
   * Register a new user with name, email, and password.
   * Note: Role is strictly determined server-side from institutional email.
   * @param {Object} credentials { name, email, password }
   * @returns {Promise<Object>} AuthResponse { token, tokenType, expiresIn, user }
   */
  async register({ name, email, password }) {
    try {
      const response = await fetch(API_ENDPOINTS.REGISTER, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Accept': 'application/json',
        },
        body: JSON.stringify({ name, email, password }),
      });

      const data = await response.json().catch(() => ({}));

      if (!response.ok) {
        const error = new Error(data.message || 'Registration failed');
        error.status = response.status;
        error.code = data.code;
        error.errors = data.errors || null;
        throw error;
      }

      return data;
    } catch (err) {
      if (err.name === 'TypeError' && err.message.includes('fetch')) {
        const netErr = new Error('Unable to connect to authentication server. Please verify the backend is running.');
        netErr.status = 0;
        throw netErr;
      }
      throw err;
    }
  },

  /**
   * Authenticate an existing user with email and password.
   * @param {Object} credentials { email, password }
   * @returns {Promise<Object>} AuthResponse { token, tokenType, expiresIn, user }
   */
  async login({ email, password }) {
    try {
      const response = await fetch(API_ENDPOINTS.LOGIN, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Accept': 'application/json',
        },
        body: JSON.stringify({ email, password }),
      });

      const data = await response.json().catch(() => ({}));

      if (!response.ok) {
        const error = new Error(data.message || 'Login failed');
        error.status = response.status;
        error.code = data.code;
        error.errors = data.errors || null;
        throw error;
      }

      return data;
    } catch (err) {
      if (err.name === 'TypeError' && err.message.includes('fetch')) {
        const netErr = new Error('Unable to connect to authentication server. Please verify the backend is running.');
        netErr.status = 0;
        throw netErr;
      }
      throw err;
    }
  },

  /**
   * Persist authenticated token and user profile in browser storage.
   * Aligns with backend stateless JWT architecture.
   */
  saveAuth(authResponse) {
    if (authResponse?.token) {
      localStorage.setItem(TOKEN_KEY, authResponse.token);
    }
    if (authResponse?.user) {
      localStorage.setItem(USER_KEY, JSON.stringify(authResponse.user));
    }
  },

  /**
   * Retrieve currently stored JWT token.
   */
  getAuthToken() {
    return localStorage.getItem(TOKEN_KEY);
  },

  /**
   * Retrieve currently stored user profile.
   */
  getAuthUser() {
    try {
      const user = localStorage.getItem(USER_KEY);
      return user ? JSON.parse(user) : null;
    } catch {
      return null;
    }
  },

  /**
   * Clear session upon logout.
   */
  clearAuth() {
    localStorage.removeItem(TOKEN_KEY);
    localStorage.removeItem(USER_KEY);
  },

  /**
   * Exchanges a single-use OAuth authorization code for an application JWT.
   * @param {string} code One-time authorization code
   * @returns {Promise<Object>} AuthResponse { token, tokenType, expiresIn, user }
   */
  async exchangeOAuthCode(code) {
    try {
      const response = await fetch(API_ENDPOINTS.OAUTH2_EXCHANGE, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Accept': 'application/json',
        },
        body: JSON.stringify({ code }),
      });

      const data = await response.json().catch(() => ({}));

      if (!response.ok) {
        const error = new Error(data.message || 'OAuth code exchange failed.');
        error.status = response.status;
        throw error;
      }

      return data;
    } catch (err) {
      if (err.name === 'TypeError' && err.message.includes('fetch')) {
        const netErr = new Error('Unable to connect to authentication server. Please verify backend is running.');
        netErr.status = 0;
        throw netErr;
      }
      throw err;
    }
  },

  /**
   * Builds Authorization headers for authenticated requests.
   */
  getAuthHeaders() {
    const token = this.getAuthToken();
    const headers = { 'Accept': 'application/json' };
    if (token) {
      headers['Authorization'] = `Bearer ${token}`;
    }
    return headers;
  },

  /**
   * Check if an active token exists.
   */
  isAuthenticated() {
    return Boolean(localStorage.getItem(TOKEN_KEY));
  },

  /**
   * Alias for getAuthUser.
   */
  getUser() {
    return this.getAuthUser();
  },

  /**
   * Updates stored user profile in local storage.
   */
  setUser(user) {
    if (user) {
      localStorage.setItem(USER_KEY, JSON.stringify(user));
    }
  },
};

